stan/math/prim/functor/apply.hpp

```diff
@@ -0,0 +1,52 @@
+#ifndef STAN_MATH_PRIM_FUNCTOR_APPLY_HPP
+#define STAN_MATH_PRIM_FUNCTOR_APPLY_HPP
+
+#include <functional>
+#include <tuple>
+#include <utility>
+
+namespace stan {
+namespace math {
+namespace internal {
+/*
+ * Invoke the functor f with arguments given in t and indexed in the index
+ * sequence I
+ *
+ * @tparam F Type of functor
+ * @tparam Tuple Type of tuple containing arguments
+ * @tparam I Parameter pack of an index sequence going from 0 to
+ * std::tuple_size<T>::value - 1 inclusive
+ * @param f functor callable
+ * @param t tuple of arguments
+ * @param i placeholder variable for index sequence
+ */
+template <class F, class Tuple, std::size_t... I>
+constexpr decltype(auto) apply_impl(F&& f, Tuple&& t,
+                                    std::index_sequence<I...> i) {
+  return f(std::forward<decltype(std::get<I>(t))>(std::get<I>(t))...);
+}
+}  // namespace internal
+/*
+ * Call the functor f with the tuple of arguments t, like:
+ *
+ * f(std::get<0>(t), std::get<1>(t), ...)
+ *
+ * TODO: replace this with implementation in C++ std when C++17 is available
+ *
+ * @tparam F Type of functor
+ * @tparam Tuple Type of tuple containing arguments
+ * @param f functor callable
+ * @param t tuple of arguments
+ */
+template <class F, class Tuple>
+constexpr decltype(auto) apply(F&& f, Tuple&& t) {
```

Comment:
We have a really similar function here: https://github.com/stan-dev/math/blob/develop/stan/math/prim/meta/index_apply.hpp
Do we need both?

Reply:
Probly not. Steve maxed out the forwarding on this one and there are some tests, so I think we go this route.

Looks like index_apply is in the kernel_generator. Do you see any problems with using apply there? If not mind swapping it out, verifying the tests for kernel generator still pass, removing index_apply, and pushing the changes here?

Reply:
I think that should work. Though I will have to undo part of refactoring in #1769. Actually I would prefer to do that in a separate PR after this and #1769 are merged - otherwise there will be a number of merge conflicts.

Reply:
Sure, seems fine to me. Can note this issue somewhere and track it?